iot-hub/Samples/device/EdgeDownstreamDevice/Program.cs

```diff
@@ -0,0 +1,131 @@
+
+using Microsoft.Azure.Devices.Client;
+
+namespace Microsoft.Azure.Devices.Client.Samples
+{
+    class Program
+    {
+
+        // 1) Obtain the connection string for your downstream device and to it
+        //    append this string GatewayHostName=<edge device hostname>;
+        // 2) The edge device hostname is the hostname set in the config.yaml of the Edge device
+        //    to which this sample will connect to.
+        //
+        // The resulting string should look like the following
+        //  "HostName=<iothub_host_name>;DeviceId=<device_id>;SharedAccessKey=<device_key>;GatewayHostName=<edge device hostname>"
+        //
+        // Either set the DEVICE_CONNECTION_STRING environment variable with this connection string
+        // or set it in the launchSettings.json.
+        private static string DeviceConnectionString = Environment.GetEnvironmentVariable("DEVICE_CONNECTION_STRING");
+
+        private static int MESSAGE_COUNT = 10;
+        private const int TEMPERATURE_THRESHOLD = 30;
+
+        private static float temperature;
+        private static float humidity;
+
+        /// <summary>
+        /// First install any CA certificate provided by the user to connect to the Edge device.
+        /// Next attempt to connect to the Edge device and send it DEVICE_MESSAGE_COUNT
+        /// number of telemetry data messages.
+        ///
+        /// Note: Either set the DEVICE_MESSAGE_COUNT environment variable with the number of
+        /// messages to be sent to the IoT Edge runtime or set it in the launchSettings.json.
+        /// </summary>
+        static void Main()
+        {
+            InstallCACert();
+
+            try
+            {
+                MESSAGE_COUNT = Int32.Parse(Environment.GetEnvironmentVariable("DEVICE_MESSAGE_COUNT"), NumberStyles.None, new CultureInfo("en-US"));
+            }
+            catch (Exception)
+            {
+                Console.WriteLine("Invalid number of messages in env variable DEVICE_MESSAGE_COUNT. MESSAGE_COUNT set to {0}\n", MESSAGE_COUNT);
+            }
+
+            DeviceClient deviceClient = DeviceClient.CreateFromConnectionString(DeviceConnectionString);
+
+            if (deviceClient == null)
+            {
+                Console.WriteLine("Failed to create DeviceClient!");
+            }
+            else
+            {
+                SendEvents(deviceClient).Wait();
+            }
+
+            Console.WriteLine("Exiting!\n");
+        }
+
+        /// <summary>
+        /// Add certificate in local cert store for use by downstream device
+        /// client for secure connection to IoT Edge runtime.
+        ///
+        ///    Note: On Windows machines, if you have not run this from an Administrator prompt,
+        ///    a prompt will likely come up to confirm the installation of the certificate.
+        ///    This usually happens the first time a certificate will be installed.
+        /// </summary>
+        static void InstallCACert()
```

Comment:
This should be an Edge API instead, not a sample.

To me, this sounds like a prerequisite: after installing edgeHub, all downstream devices will need to trust it's certificates. IoT developers will need to follow SDL to set up scheduled rotation of such certificates:

  a. Issue edge certificates from public CAs (that are implicitly trusted by some of the platforms)
  b. Issue edge certificates from private CAs and ensure that proper PKI is followed (cert rotation, good expiration dates - i.e. 3-6months, etc)